From 906bef5d54b7bfac8d86632b1741076bd152932a Mon Sep 17 00:00:00 2001
From: Andrew Branson <andrew.branson@jolla.com>
Date: Wed, 24 Jun 2020 07:56:50 +0200
Subject: [PATCH] [sailfishos][gecko] Prioritize GMP plugins over all others,
 and support decoding video for h264, vp8 & vp9.

Automatically load gmp-droid at startup

Signed-off-by: Raine Makelainen <raine.makelainen@jolla.com>
---
 dom/media/platforms/PDMFactory.cpp               | 13 +++++++------
 .../platforms/agnostic/gmp/GMPDecoderModule.cpp  | 16 ++++++++++++++++
 .../mozapps/extensions/internal/GMPProvider.jsm  | 15 +++++++++++++++
 3 files changed, 38 insertions(+), 6 deletions(-)

diff --git a/dom/media/platforms/PDMFactory.cpp b/dom/media/platforms/PDMFactory.cpp
index c8e7df236d1b..d47ec17b42af 100644
--- a/dom/media/platforms/PDMFactory.cpp
+++ b/dom/media/platforms/PDMFactory.cpp
@@ -316,6 +316,12 @@ void PDMFactory::CreatePDMs() {
     return;
   }
 
+  if (MediaPrefs::PDMGMPEnabled()) {
+    m = new GMPDecoderModule();
+    mGMPPDMFailedToStartup = !StartupPDM(m);
+  } else {
+    mGMPPDMFailedToStartup = false;
+  }
 #ifdef XP_WIN
   if (MediaPrefs::PDMWMFEnabled() && !IsWin7AndPre2000Compatible()) {
     m = new WMFDecoderModule();
@@ -354,12 +360,7 @@ void PDMFactory::CreatePDMs() {
   m = new AgnosticDecoderModule();
   StartupPDM(m);
 
-  if (MediaPrefs::PDMGMPEnabled()) {
-    m = new GMPDecoderModule();
-    mGMPPDMFailedToStartup = !StartupPDM(m);
-  } else {
-    mGMPPDMFailedToStartup = false;
-  }
+
 }
 
 void PDMFactory::CreateNullPDM() {
diff --git a/dom/media/platforms/agnostic/gmp/GMPDecoderModule.cpp b/dom/media/platforms/agnostic/gmp/GMPDecoderModule.cpp
index 7f7fb437ee0a..291f0334767c 100644
--- a/dom/media/platforms/agnostic/gmp/GMPDecoderModule.cpp
+++ b/dom/media/platforms/agnostic/gmp/GMPDecoderModule.cpp
@@ -88,6 +88,22 @@ bool GMPDecoderModule::SupportsMimeType(const nsACString& aMimeType,
 
 bool GMPDecoderModule::SupportsMimeType(
     const nsACString& aMimeType, DecoderDoctorDiagnostics* aDiagnostics) const {
+#if defined(MOZ_FMP4)
+  if (MP4Decoder::IsH264(aMimeType)) {
+    return HaveGMPFor(NS_LITERAL_CSTRING(GMP_API_VIDEO_DECODER),
+                      { NS_LITERAL_CSTRING("h264") });
+  }
+#endif
+
+  if (VPXDecoder::IsVP9(aMimeType)) {
+    return HaveGMPFor(NS_LITERAL_CSTRING(GMP_API_VIDEO_DECODER),
+                      { NS_LITERAL_CSTRING("vp9") });
+  }
+
+  if (VPXDecoder::IsVP8(aMimeType)) {
+    return HaveGMPFor(NS_LITERAL_CSTRING(GMP_API_VIDEO_DECODER),
+                      { NS_LITERAL_CSTRING("vp8") });
+  }
   return false;
 }
 
diff --git a/toolkit/mozapps/extensions/internal/GMPProvider.jsm b/toolkit/mozapps/extensions/internal/GMPProvider.jsm
index 4331b11da9c2..0b58a5e96319 100644
--- a/toolkit/mozapps/extensions/internal/GMPProvider.jsm
+++ b/toolkit/mozapps/extensions/internal/GMPProvider.jsm
@@ -32,6 +32,8 @@ const GMP_CHECK_DELAY        = 10 * 1000; // milliseconds
 const NS_GRE_DIR             = "GreD";
 const CLEARKEY_PLUGIN_ID     = "gmp-clearkey";
 const CLEARKEY_VERSION       = "0.1";
+const DROID_PLUGIN_ID        = "gmp-droid";
+const DROID_VERSION          = "0.1";
 
 const GMP_LICENSE_INFO       = "gmp_license_info";
 const GMP_PRIVACY_INFO       = "gmp_privacy_info";
@@ -573,6 +575,19 @@ var GMPProvider = {
     } catch (e) {
       this._log.warn("startup - adding clearkey CDM failed", e);
     }
+
+    try {
+      let greDir = Services.dirsvc.get(NS_GRE_DIR,
+                                       Ci.nsILocalFile);
+      let droidPath = OS.Path.join(greDir.path,
+                                      DROID_PLUGIN_ID,
+                                      DROID_VERSION);
+      this._log.info("startup - adding droidmedia GMP directory " +
+                     droidPath);
+      gmpService.addPluginDirectory(droidPath);
+    } catch (e) {
+      this._log.warn("startup - adding droidmedia GMP failed", e);
+    }
   },
 
   shutdown() {
-- 
2.26.2

